File: 525-Contiguous Array.py
class Solution(object):
    def findMaxLength(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
# https://www.youtube.com/watch?v=uAGt1QoAoMU
# if nums[i] == 0, sum = sum-1, sum_dict[sum] = i, else sum = sum+1, sum_dict[sum] = i

# time complexity: O(n)
        
        res = 0
        hmap = {0: -1}
        _sum = 0
        for i in range(len(nums)):
            if nums[i] == 0:
                _sum -= 1
            else:
                _sum += 1
            
            if _sum in hmap:
                res = max(res, i-hmap[_sum])
            else:
                hmap[_sum] = i
        
        return res
        

File: C++/525-Contiguous Array.cpp
class Solution {
public:
    int findMaxLength(vector<int>& nums) {
        unordered_map<int, int> preSum;
        preSum[0] = -1;
        int curr = 0, res = 0;
        
        for (int i=0; i<nums.size(); i++) {
            curr += nums[i] == 1?1:-1;
            if (preSum.count(curr)) {
                res = max(res, i-preSum[curr]);
            } else {
                preSum[curr] = i;
            }
        }
        
        return res;
    }
};